### ISSUE TITLE
tf.keras.optimizers.experimental.AdamW only support constant weight_decay

### ISSUE THREAD

**Issue by x10000year** [keras-team/keras#16517]
tf.keras.optimizers.experimental.AdamW only support constant weight_decay. The legacy tfa.optimizers.AdamW supports callable weight_decay, which is much better.

```
model = MyModel()
schedule = tf.keras.optimizers.schedules.PolynomialDecay(
    1.0,
    decay_steps=10000)
learning_rate_fn = lambda: learning_rate * schedule(model.iterations.read_value())
weight_decay_fn = lambda: weight_decay * schedule(model.iterations.read_value())
optimizer = tfa.optimizers.AdamW(weight_decay=weight_decay_fn,
                                 learning_rate=learning_rate_fn)
```

As above code shows, tfa.optimizers.AdamW allows us to specify the schedule of weight decay, which should be in proportional with learning rate schedule. However, I cannot do the same thing with tf.keras.optimizers.experimental.AdamW.

**Comment by sushreebarsa** [comment]
@x10000year 
Please elaborate about your feature and  specify the use cases for this feature. Thank you!

**Comment by x10000year** [comment]
The feature requested is to support dynamic weight decay in tf.keras.optimizers.experimental.AdamW, so that weight decay value and learning rate can use similar schedule. This is important because for most learning rate schedule learning rate can decays to very small or zero value. When learning rate becomes close to zero, we should not use the same weight decay which we use when the learning rate is big.

**Comment by chenmoneygithub** [comment]
@x10000year Thanks for reporting the issue! 

sorry for the late reply, thinking about this - do we really need to let weight decay follow the learning rate schedule? My thought is when the learning rate is small, the value change to model variables is small as well. Hence even if we keep the same weight decay rate, the weight decay's effect is getting smaller. 

Taking Pytorch's doc for reference (they have the updating rule algo, which we should possibly have too...), when learning rate becomes small, the decay becomes less. 

Let me know your thoughts, thanks!

**Comment by lucasdavid** [comment]
@x10000year: I apologize if you're already ahead of me on this and I didn't understand what you wrote, but just to make sure we are on the same page: differently than Pytorch's [torch.optim._single_tensor_sgd#L224-L241](https://github.com/pytorch/pytorch/blob/master/torch/optim/sgd.py#L224-L241) and [keras.optimizers.experimental.AdamW#L173](https://github.com/keras-team/keras/blob/v2.9.0/keras/optimizers/optimizer_experimental/adamw.py#L173), tensorflow-addons' implementation does not multiply `wd` by `lr`, so a constant parameter renders a constant factor throughout the training. On the other hand, `experimental.AdamW` considers decay as `param -= lr*(1-wd)*param`. Hence a constant parameter is still reduced throughout training if the learning rate is decayed as well.

I suppose one argument could be made that this implementation isn't as flexible as tfa's, and that you might find models or articles that use different strategies for learning rate decay and weight decay.

@chenmoneygithub: "Hence even if we keep the same weight decay rate, the weight decay's effect is getting smaller." -- true, but that's also true for the gradient vectors' norms. When the learning rate is decayed over time, `lr*grad*param` tends to 0 faster than `wd*param`, and the relative importance of the decay becomes overwhelming.
I have a couple of training stats that exemplify how training diverges if `lr` is reduced and `wd` is kept constant (using `tfa.optimizers.SGDW`), if you want.

**Comment by lucasdavid** [comment]
Oh, I see the implementation has been changed in commit [0a43b88](https://github.com/keras-team/keras/commit/0a43b8818293e4b458a645b86f70129fa12b6062) and no longer multiplies wd by the learning rate. I strongly believe this is a mistake and I hope Keras dev reconsider this.

**Comment by chenmoneygithub** [comment]
Sorry, I forgot to reply. We have fixed the multiplier issue.
